### PR TITLE
Process arguments better when querying posts for automated latest content [MAILPOET-4082]

### PR DESCRIPTION
**Pull request by samnajian**
Fixes [MAILPOET-4082]

[MAILPOET-4082]: https://mailpoet.atlassian.net/browse/MAILPOET-4082?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: mailpoet/lib/Newsletter/AutomatedLatestContent.php [hunk 1]
```diff
@@ -65,8 +65,13 @@ public function getPosts($args, $postsToExclude = [], $newsletterId = false, $ne
       'post_type' => (isset($args['contentType'])) ? $args['contentType'] : 'post',
       'post_status' => (isset($args['postStatus'])) ? $args['postStatus'] : 'publish',
       'orderby' => 'date',
-      'order' => ($args['sortBy'] === 'newest') ? 'DESC' : 'ASC',
+      'order' => isset($args['sortBy']) && in_array($args['sortBy'], ['newest', 'DESC']) ? 'DESC' : 'ASC',
     ];
+
+    if (!$this->wp->isPostTypeViewable($parameters['post_type'])) {
+      throw new \InvalidArgumentException($this->wp->__('Invalid request parameters', 'mailpoet'));
```
Comment (JanJakes) [inline mailpoet/lib/Newsletter/AutomatedLatestContent.php:75]: I think we should throw here one of the [exceptions in the MailPoet namespace](https://github.com/mailpoet/mailpoet/blob/master/mailpoet%2Flib%2Fexceptions.php). Picking the correct one will also result in the correct HTTP code on the API (which I think should be 4xx, in this case). Now, this would probably result in error 500, which is not desired.

I'd suggest using this one:
```php
throw AccessDeniedException::create()
  ->withErrors([APIError::BAD_REQUEST => __('...', 'mailpoet')]);
```

Also the message could reflect that it's about access denied, I guess.

File: mailpoet/lib/WP/Functions.php [hunk 2]
```diff
@@ -776,4 +776,8 @@ public function escUrlRaw(string $url, array $protocols = null): string {
   public function restUrl(string $path = '', string $scheme = 'rest'): string {
     return rest_url($path, $scheme);
   }
+
+  public function isPostTypeViewable(string $post_type): bool {
```
Comment (JanJakes) [inline mailpoet/lib/WP/Functions.php:780]: This is a tiny issue but the argument should be camel-cased (`$postType`).

File: mailpoet/lib/WP/Functions.php [hunk 3]
```diff
@@ -780,4 +780,8 @@ public function restUrl(string $path = '', string $scheme = 'rest'): string {
   public function isPostTypeViewable(string $post_type): bool {
     return is_post_type_viewable($post_type);
   }
+
+  public function isPostStatusViewable(string $post_status): bool {
```
Comment (JanJakes) [inline mailpoet/lib/WP/Functions.php:784]: Same here (camel case).

File: mailpoet/tests/integration/Newsletter/Blocks/AbandonedCartContentTest.php [hunk 4]
```diff
@@ -78,6 +78,14 @@ public function _before() {
       $this->wp->wpDeletePost((int)$product->ID);
     }
 
+
+    if (class_exists('\WC_Post_Types')) {
+      \WC_Post_Types::register_post_types();
```
Comment (JanJakes) [inline mailpoet/tests/integration/Newsletter/Blocks/AbandonedCartContentTest.php:83]: I think we don't have WC in integration tests active at the moment so I'd remove the IF part and leave just the `else` branch now. What do you think?

File: mailpoet/lib/Newsletter/AutomatedLatestContent.php [hunk 5]
```diff
@@ -65,8 +65,13 @@ public function getPosts($args, $postsToExclude = [], $newsletterId = false, $ne
       'post_type' => (isset($args['contentType'])) ? $args['contentType'] : 'post',
       'post_status' => (isset($args['postStatus'])) ? $args['postStatus'] : 'publish',
       'orderby' => 'date',
-      'order' => ($args['sortBy'] === 'newest') ? 'DESC' : 'ASC',
+      'order' => isset($args['sortBy']) && in_array($args['sortBy'], ['newest', 'DESC']) ? 'DESC' : 'ASC',
```
Comment (JanJakes) [inline mailpoet/lib/Newsletter/AutomatedLatestContent.php:68]: Was this broken? I don't know what the order params look like but is this a fix? It's probably something I'd normally extract to another commit to describe the fix in the commit message but of course no need to do it now.

**Comment by JanJakes** [comment]
I'm also adding @websupporter for a review here as he was the one who reported this issue. Pls @websupporter, could you have a look if the added checks look good to you? Thanks!

File: mailpoet/lib/Newsletter/AutomatedLatestContent.php (hunk 5 above)
Comment (samnajian) [reply mailpoet/lib/Newsletter/AutomatedLatestContent.php:68]: It was not broken per se, but it was wrong, it could easily force `DESC` to be interpreted as `ASC` and besides not checking if `sortBy` was set, I can put it to another commit

File: mailpoet/tests/integration/Newsletter/Blocks/AbandonedCartContentTest.php (hunk 4 above)
Comment (samnajian) [reply mailpoet/tests/integration/Newsletter/Blocks/AbandonedCartContentTest.php:83]: Having it here doesn't do any harms, and for now it also implies that we're better off having WC active but it may create benefits in future, I guess some of the flakey tests come from not using the correct functions to register stuff like this, this is only a guess though

Comment (JanJakes) [reply mailpoet/tests/integration/Newsletter/Blocks/AbandonedCartContentTest.php:83]: @samnajian Right, but on the other hand, I don't think it's a good practice to have different possible flows in tests (if/else based on some circumstances). Again, WC is currently not there so it doesn't make sense to have an IF "in case it is there". Also (noticing just now), the post types should be removed in `cleanup`.

Comment (samnajian) [reply mailpoet/tests/integration/Newsletter/Blocks/AbandonedCartContentTest.php:83]: I see your point, fine removing the other branch :)

**Review by JanJakes — APPROVED** [review]

**Comment by websupporter** [comment]
I am not sure, if I would add those checks to `Newsletter/AutomatedLatestContent`. Imo the classes responsibility is to return posts based on query arguments. `API/JSON/v1/AutomatedLatestContent` on the other hand is responsible of returning this in a specific context. I think the checks should be there.

The current implementation breaks some features:
The endpoint is used to fetch drafts/scheduled and private posts in the Email editor:
image
image

Imo that makes sense: I draft a post. I draft a newsletter. I publish post. I send newsletter. Same for schedule and also private. So the evaluation should imo be on a user basis and not on the post status basis.

File: mailpoet/lib/API/JSON/v1/AutomatedLatestContent.php [hunk 6]
```diff
@@ -65,14 +75,24 @@ public function getTerms($data = []) {
     return $this->successResponse(array_values($terms));
   }
 
+  /**
+   * @param int[]|\WP_Post[] $posts
+   * @return \WP_Post[]
+   */
+  private function getPermittedPosts($posts) {
+    return array_filter($posts, function ($post) {
+      return !is_numeric($post) && $this->permissionHelper->checkReadPermission($post);
```
Comment (websupporter) [inline mailpoet/lib/API/JSON/v1/AutomatedLatestContent.php:84]: Why `!is_numeric()`? I mean, why expect numeric values at all? I know `get_posts()` can return `int[]` with `["fields" => "ids"]` which we don't use though. Maybe we could make sure, this does not happen for posts returned from ACL, so that we do not have this ambiguity trickling down in our code base?

Comment (samnajian) [reply mailpoet/lib/API/JSON/v1/AutomatedLatestContent.php:84]: @websupporter since we were are not preventing the int[] output, this is needed per type definitions! not sure if we want to prevent this behaviour now, I can create a Jira ticket to check this in depth and probably refactor, what do you think?

**Review by websupporter — DISMISSED** [review]
lgtm

File: mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        return $post;
+      },
+    ]);
+    $apiPermissionHelper = $this->diContainer->get(APIPermissionHelper::class);
+    $this->alcAPI = new ALCAPI($alc, $apiPermissionHelper, $this->wp);
+
+  }
+
+  public function tearDown(): void {
+    parent::tearDown();
```
Comment (JanJakes) [inline mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php:103]: Is there a difference from using `_after`? I don't remember now...

File: mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+  }
+
+  private function loginWithRole(string $role): void {
+    $username = uniqid("testUser");
+    $email = "$username@test.com";
+    $existingUser = $this->wp->getUserBy("email", $email);
+
+    if ($existingUser) {
+      wp_delete_user($existingUser->ID);
+    }
```
Comment (JanJakes) [inline mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php:119]: Maybe this could be in before/after cleanups?

File: mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    ];
+    $draftPost = [
+      'post_title' => 'Draft Post',
+      'post_content' => 'contents',
+      'post_status' => 'draft',
+    ];
+
+    array_map(function ($postarr){
+      wp_insert_post($postarr);
+    }, [$privatePost, $publicPost, $draftPost]);
```
Comment (JanJakes) [inline mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php:206]: Running `$privatePost = wp_insert_post([...])` directly would make it more consistent with the other test and would allow to get rid of this `array_map`.

File: mailpoet/lib/Util/APIPermissionHelper.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace MailPoet\Util;
+
+// \WP_REST_Posts_Controller is always available in WP runtime but not when building the plugin outside of WP env
+if (class_exists("\WP_REST_Posts_Controller")) {
```
[the remark this answers is not in the document]
Comment (samnajian) [reply mailpoet/lib/Util/APIPermissionHelper.php:6]: > Could we maybe do this (see below) instead of the conditional class definition?
> 
> ```
> require_once ABSPATH . '/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php';
> ```
> 
> We [already have a couple of these](https://github.com/mailpoet/mailpoet/search?q=require_once+ABSPATH). I'm curious how DI dump actually works when the class is only conditional 🤔

I actually tried the `require_once` approach first, but PHPStan was not happy with it, this if statement approach works and I've verified the `FreeCachedContainer ` has `APIPermissionHelper ` in the build artifact, but still would be happy to learn if you think this approach is error prone

File: mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php (hunk 7 above)
Comment (samnajian) [reply mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php:103]: tearDown is after and `_before` is well before, na?

File: mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php (hunk 8 above)
Comment (samnajian) [reply mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php:119]: No, this is done to make sure this user is not existing before trying to add it, but I'll add user cleanup anyway

File: mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php (hunk 7 above)
Comment (JanJakes) [reply mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php:103]: I'm not sure I understand your answer but my question could be reformulated as "should we maybe use `_after` here?".

File: mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php (hunk 9 above)
Comment (samnajian) [reply mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php:206]: 1) In this test I need access to `post_title` of all 3 created posts, removing these variables would mean I'll need to create another 3 variables which doesn't make any difference .
2) I'm not really sure what value  `consistency between different tests` can bring to us, nor do I know how getting rid of `array_map` can serve us

File: mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php (hunk 8 above)
Comment (JanJakes) [reply mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php:119]: > No
Well, it can be in `_before` but I guess the username needs to be fixed and not random. It was a rhetorical question as we don't usually do these things in the tests themselves.

File: mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php (hunk 9 above)
Comment (JanJakes) [reply mailpoet/tests/integration/Newsletter/AutomatedLatestContentAPITest.php:204]: 1. Right, that explains it.
2. Not sure about the need for such a comment, I'm only trying to simplify code.

**Review by websupporter — DISMISSED** [review]
lgtm

**Review by JanJakes — APPROVED** [review]

**Review by veljkho — APPROVED** [review]
